File: skill-sim-project/backend/src/main/java/com/example/skillsim/dto/MethodologyResponse.java
package com.example.skillsim.dto;

import java.util.List;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class MethodologyResponse {
    private FormulaInfo formula;
    private Map<String, Double> statWeights;
    private ConditionProbabilitiesInfo conditionProbabilities;

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class FormulaInfo {
        private FormulaItem perSkillFormula;
        private FormulaItem totalFormula;
        private FormulaItem percentEffectRule;
        private FormulaItem roundingRule;
        private FormulaItem conditionCombinationRule;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class FormulaItem {
        private String displayText;
        private String descriptionKey;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ConditionProbabilitiesInfo {
        private List<ConditionProbabilityEntry> staticProbabilities;
        private List<RoleProbabilityEntry> roleProbabilities;
        private List<ConditionProbabilityEntry> battingOrderProbabilities;
        private List<ReachProbabilityEntry> reachProbabilities;
        private List<GateEntry> gates;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ConditionProbabilityEntry {
        private String token;
        private double value;
        private String descriptionKey;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class RoleProbabilityEntry {
        private String role;
        private List<Double> inningWeights;
        private double gutsProbability;
        private double nineBatterDuration;
        private double maestroCumulative;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ReachProbabilityEntry {
        private String orderGroup;
        private String descriptionKey;
        private List<Double> reachProbabilities;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GateEntry {
        private String token;
        private String descriptionKey;
    }
}

File: backend/src/main/kotlin/com/example/skillsim/dto/MethodologyResponse.kt
package com.example.skillsim.dto

data class MethodologyResponse(
    val formula: FormulaInfo,
    val statWeights: Map<String, Double>,
    val conditionProbabilities: ConditionProbabilitiesInfo,
) {
    data class FormulaInfo(
        val perSkillFormula: FormulaItem,
        val totalFormula: FormulaItem,
        val percentEffectRule: FormulaItem,
        val roundingRule: FormulaItem,
        val conditionCombinationRule: FormulaItem,
    )

    data class FormulaItem(
        val displayText: String,
        val descriptionKey: String,
    )

    data class ConditionProbabilitiesInfo(
        val staticProbabilities: List<ConditionProbabilityEntry>,
        val roleProbabilities: List<RoleProbabilityEntry>,
        val battingOrderProbabilities: List<ConditionProbabilityEntry>,
        val reachProbabilities: List<ReachProbabilityEntry>,
        val gates: List<GateEntry>,
    )

    data class ConditionProbabilityEntry(
        val token: String,
        val value: Double,
        val descriptionKey: String,
    )

    data class RoleProbabilityEntry(
        val role: String,
        val inningWeights: List<Double>,
        val gutsProbability: Double,
        val patienceBelowVelocityProbability: Double,
        val nineBatterDuration: Double,
        val maestroCumulative: Double,
    )

    data class ReachProbabilityEntry(
        val orderGroup: String,
        val descriptionKey: String,
        val reachProbabilities: List<Double>,
    )

    data class GateEntry(
        val token: String,
        val descriptionKey: String,
    )
}
